govcd/vm_affinity_rule.go

```diff
@@ -0,0 +1,397 @@
+// If polarity is indicated, the function retrieves only the rules with the given polarity
+func (vdc *Vdc) GetVmAffinityRulesByName(name string, polarity string) ([]*VmAffinityRule, error) {
+
+	var returnList []*VmAffinityRule
+	ruleList, err := vdc.GetAllVmAffinityRuleList()
+	if err != nil {
+		return nil, err
+	}
+	for _, rule := range ruleList {
+		if rule.Name == name {
+			fullRule, err := vdc.GetVmAffinityRuleByHref(rule.HREF)
+			if err != nil {
+				return returnList, err
+			}
+			if (polarity != "" && polarity == rule.Polarity) || polarity == "" {
+				returnList = append(returnList, fullRule)
+			}
+		}
+	}
+	return returnList, nil
+}
+
+// GetVmAffinityRuleById retrieves a VM affinity or anti-affinity rule by ID
+func (vdc *Vdc) GetVmAffinityRuleById(id string) (*VmAffinityRule, error) {
+
+	list, err := vdc.GetAllVmAffinityRuleList()
+	if err != nil {
+		return nil, err
+	}
+	for _, rule := range list {
+		if equalIds(id, rule.ID, rule.HREF) {
+			return vdc.GetVmAffinityRuleByHref(rule.HREF)
+		}
+	}
+	return nil, ErrorEntityNotFound
+}
+
+// GetVmAffinityRuleByNameOrId retrieves an affinity or anti-affinity rule by name or ID
+// Given the possibility of a name identifying multiple items, this function may also fail
+// when the search by name returns more than one item.
+func (vdc *Vdc) GetVmAffinityRuleByNameOrId(identifier string) (*VmAffinityRule, error) {
+	getByName := func(name string, refresh bool) (interface{}, error) {
+		list, err := vdc.GetVmAffinityRulesByName(name, "")
+		if err != nil {
+			return nil, err
+		}
+		if len(list) == 0 {
+			return nil, ErrorEntityNotFound
+		}
+		if len(list) == 1 {
+			return list[0], nil
+		}
+		return nil, fmt.Errorf("more than item matches the name '%s'", name)
+	}
+	getById := func(id string, refresh bool) (interface{}, error) { return vdc.GetVmAffinityRuleById(id) }
+	entity, err := getEntityByNameOrId(getByName, getById, identifier, false)
+	if entity == nil {
+		return nil, err
+	}
+	return entity.(*VmAffinityRule), err
+}
+
+func validateAffinityRule(client *Client, affinityRuleDef *types.VmAffinityRule, checkVMs bool) (*types.VmAffinityRule, error) {
+	if affinityRuleDef == nil {
+		return nil, fmt.Errorf("empty definition given for a VM affinity rule")
+	}
+	if affinityRuleDef.Name == "" {
+		return nil, fmt.Errorf("no name given for a VM affinity rule")
+	}
+	if affinityRuleDef.Polarity == "" {
+		return nil, fmt.Errorf("no polarity given for a VM affinity rule")
+	}
+	if !correctPolarity(affinityRuleDef.Polarity) {
+		return nil, fmt.Errorf("illegal polarity given (%s) for a VM affinity rule", affinityRuleDef.Polarity)
+	}
+	//if len(affinityRuleDef.VmReferences) == 0 || len (affinityRuleDef.VmReferences[0].VMReference) <2 {
+	//	return nil, fmt.Errorf("at least 2 VMs should be given for a VM Affinity Rule")
+	//}
+	// Ensure the VMs in the list are different
+	var seenVms = make(map[string]bool)
```

Comment:
If I haven't given `demo` I would guess what this do and why this need. At least comment what this part aim to do is needed.

Reply:
I have improved comments in this code

Reply:
thank you!